File: ds_algo/grind75/704_BinarySearch.py
### GRIND 75 SECOND PASS ###
"""
approach:
    binary search:
        set left and right as 0 and len-1 index
        find the middle
        compare the middle to the target
            if equal, return middle index
            if target is greater, store middle index in left
            if target is smaller, store middle index in right
            repeat until left == right
    
    bisect (convenience function):
        from the forum.
        use bisect to determine at which index target would need to be inserted to maintain sorted order
        if the index is within the bounds of the existing list (i.e. target is not > or < the entire list)
            then return the index only if nums[index] is equal to the target
            otherwise, return -1 because the target does not exist in the list
"""
# bisect (convenience function)
# T: O(logn) assuming that bisect implements a binary search
# S: O(1)
class Solution:
    def search(self, nums, target):
        index = bisect.bisect_left(nums, target)
        return index if index < len(nums) and nums[index] == target else -1

# binary search
# T: O(logn)
# S: O(1)
class __Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums)-1
        while left <= right:
            middle = (left+right)//2
            if target == nums[middle]:
                return middle
            elif target > nums[middle]:
                left = middle + 1
            else:
                right = middle - 1
        return -1

File: ds_algo/grind75/704_BinarySearch.cpp
// class Solution {
// public:
//     // standard approach4
//     // O(logn) because each operation cuts the array in 2 and we run until there's no array left
//     int search(vector<int>& nums, int target) {
//         int upper = nums.size()-1;
//         int lower = 0;
//         int i = 0;

//         // if target is less than the first index or greater than the last, it's not in the list
//         if((target < nums[0]) || (target > nums[upper])) {
//             return -1;
//         }
//         // check if upper or lower are target; helps with the edge case where array length is 2
//         if(target == nums[lower]) { return 0; }
//         if(target == nums[upper]) { return upper; }

//         while(true) {
//             i = (lower + upper)/2;

//             // return the index if we found a match
//             if(target == nums[i]) { return i; }
//             // if lower and upper are adjacent, we checked everything so return
//             if(upper - lower == 1) { return -1; }

//             // adjust the bounds if no match
//             target > nums[i] ? lower = i : upper = i;
//         }
//         return -1;
//     }
// };
